---
title: "GeneScoreMatrix"
output: html_notebook
---

# Modify the .addGeneScoreMat function

Disable library size normalization to `scaleTo`.
Instead, we use Linnorm.

```{r}
.myaddGeneScoreMat <- function(
  i = NULL,
  ArrowFiles = NULL,
  genes = NULL,
  geneModel = "exp(-abs(x)/5000) + exp(-1)",
  matrixName = "GeneScoreMatrix",
  extendUpstream = c(1000, 100000),
  extendDownstream = c(1000, 100000),
  geneUpstream = 5000, #New Param
  geneDownstream = 0, #New Param
  useGeneBoundaries = TRUE,
  useTSS = FALSE, #New Param
  extendTSS = FALSE,
  tileSize = 500,
  ceiling = 4,
  geneScaleFactor = 5, #New Param
  scaleTo = 10000,
  excludeChr = c("chrY","chrM"),
  blacklist = NULL,
  cellNames = NULL,
  allCells = NULL,
  force = FALSE,
  tmpFile = NULL,
  subThreads = 1,
  tstart = NULL,
  logFile = NULL
  ){

  ArchR:::.validInput(input = i, name = "i", valid = c("integer"))
  ArchR:::.validInput(input = ArrowFiles, name = "ArrowFiles", valid = c("character"))
  ArchR:::.validInput(input = genes, name = "genes", valid = c("GRanges"))
  ArchR:::.validInput(input = geneModel, name = "geneModel", valid = c("character"))
  ArchR:::.validInput(input = matrixName, name = "matrixName", valid = c("character"))
  ArchR:::.validInput(input = extendUpstream, name = "extendUpstream", valid = c("integer"))
  ArchR:::.validInput(input = extendDownstream, name = "extendDownstream", valid = c("integer"))
  ArchR:::.validInput(input = tileSize, name = "tileSize", valid = c("integer"))
  ArchR:::.validInput(input = ceiling, name = "ceiling", valid = c("integer"))
  ArchR:::.validInput(input = useGeneBoundaries, name = "useGeneBoundaries", valid = c("boolean"))
  ArchR:::.validInput(input = scaleTo, name = "scaleTo", valid = c("numeric"))
  ArchR:::.validInput(input = excludeChr, name = "excludeChr", valid = c("character", "null"))
  ArchR:::.validInput(input = blacklist, name = "blacklist", valid = c("GRanges", "null"))
  ArchR:::.validInput(input = cellNames, name = "cellNames", valid = c("character", "null"))
  ArchR:::.validInput(input = allCells, name = "allCells", valid = c("character", "null"))
  ArchR:::.validInput(input = force, name = "force", valid = c("boolean"))
  ArchR:::.validInput(input = tmpFile, name = "tmpFile", valid = c("character", "null"))

  if(inherits(mcols(genes)$symbol, "list") | inherits(mcols(genes)$symbol, "SimpleList")){
    stop("Found a list in genes symbol! This is an incorrect format. Please correct your genes!")
  }

  ArrowFile <- ArrowFiles[i]
  sampleName <- ArchR:::.sampleName(ArrowFile)

  if(is.null(tmpFile)){
    tmpFile <- ArchR:::.tempfile(pattern = paste0("tmp-", ArchR:::.sampleName(ArrowFile)))
  }

  #Check
  o <- h5closeAll()
  o <- ArchR:::.createArrowGroup(ArrowFile = ArrowFile, group = matrixName, force = force, logFile = logFile)

  geneRegions <- genes[BiocGenerics::which(seqnames(genes) %bcni% excludeChr)]
  seqlevels(geneRegions) <- as.character(unique(seqnames(geneRegions)))
  geneRegions <- geneRegions[!is.na(mcols(geneRegions)$symbol)]

  #Create Gene Regions Then Remove Strand Column
  if(useTSS){
    ArchR:::.logMessage(paste0(sampleName, " .addGeneScoreMat useTSS = TRUE"))
    distMethod <- "GenePromoter"
    geneRegions$geneStart <- start(resize(geneRegions, 1, "start"))
    geneRegions$geneEnd <- start(resize(geneRegions, 1, "end"))
    geneRegions <- resize(geneRegions, 1, "start")
    if(extendTSS){
      geneRegions <- extendGR(gr = geneRegions, upstream = geneUpstream, downstream = geneDownstream)
    }
    geneRegions$geneWeight <- geneScaleFactor
  }else{
    ArchR:::.logMessage(paste0(sampleName, " .addGeneScoreMat useTSS = FALSE"))
    distMethod <- "GeneBody"
    geneRegions$geneStart <- start(resize(geneRegions, 1, "start"))
    geneRegions$geneEnd <- start(resize(geneRegions, 1, "end"))
    geneRegions <- extendGR(gr = geneRegions, upstream = geneUpstream, downstream = geneDownstream)
    m <- 1 / width(geneRegions)
    geneRegions$geneWeight <- 1 + m * (geneScaleFactor - 1) / (max(m) - min(m))
  }

  ArchR:::.logDiffTime(sprintf("Computing Gene Scores using distance relative to %s! ", distMethod), tstart, logFile = logFile)

  #Add Gene Index For ArrowFile
  geneRegions <- sort(sortSeqlevels(geneRegions), ignore.strand = TRUE)
  ArchR:::.logThis(geneRegions, paste0(sampleName, " .addGeneScoreMat geneRegions"), logFile = logFile)
  
  geneRegions <- split(geneRegions, seqnames(geneRegions))
  geneRegions <- lapply(geneRegions, function(x){
    mcols(x)$idx <- seq_along(x)
    return(x)
  })

  #Blacklist Split
  if(!is.null(blacklist)){
    if(length(blacklist) > 0){
      blacklist <- split(blacklist, seqnames(blacklist))
    }
  }

  #Get all cell ids before constructing matrix
  if(is.null(cellNames)){
    cellNames <- ArchR:::.availableCells(ArrowFile)
  }

  if(!is.null(allCells)){
    cellNames <- cellNames[cellNames %in% allCells]
  }

  tstart <- Sys.time()


  #########################################################################################################
  #First we will write gene scores to a temporary path! rhdf5 delete doesnt actually delete the memory!
  #########################################################################################################
  totalGS <- ArchR:::.safelapply(seq_along(geneRegions), function(z){

    totalGSz <- tryCatch({

      ArchR:::.logDiffTime(sprintf("Creating Temp GeneScoreMatrix for %s, Chr (%s of %s)!", sampleName, z, length(geneRegions)), 
        tstart, verbose = FALSE, logFile = logFile)

      #Get Gene Starts
      geneRegionz <- geneRegions[[z]]
      geneRegionz <- geneRegionz[order(geneRegionz$idx)]
      chrz <- paste0(unique(seqnames(geneRegionz)))

      #Read in Fragments
      frag <- ArchR:::.getFragsFromArrow(ArrowFile, chr = chrz, out = "IRanges", cellNames = cellNames)
      fragSt <- trunc(start(frag)/tileSize) * tileSize
      fragEd <- trunc(end(frag)/tileSize) * tileSize
      fragBC <- rep(S4Vectors::match(mcols(frag)$RG, cellNames), 2)
      rm(frag)
      gc()

      #Unique Inserts
      uniqIns <- sort(unique(c(fragSt,fragEd)))

      #Construct tile by cell mat!
      matGS <- Matrix::sparseMatrix(
          i = match(c(fragSt, fragEd), uniqIns),
          j = as.vector(fragBC),
          x = rep(1,  2*length(fragSt)),
          dims = c(length(uniqIns), length(cellNames))
        )  
      
      if(!is.null(ceiling)){
        matGS@x[matGS@x > ceiling] <- ceiling
      }

      #Unique Tiles
      uniqueTiles <- IRanges(start = uniqIns, width = tileSize)
      
      #Clean Memory
      rm(uniqIns, fragSt, fragEd, fragBC)
      gc() 

      #Time to Overlap Gene Windows
      if(useGeneBoundaries){

        geneStartz <- start(resize(geneRegionz, 1, "start"))
        geneEndz <- start(resize(geneRegionz, 1, "end"))

        pminGene <- pmin(geneStartz, geneEndz)
        pmaxGene <- pmax(geneStartz, geneEndz)

        idxMinus <- BiocGenerics::which(strand(geneRegionz) != "-")
    
        pReverse <- rep(max(extendDownstream), length(pminGene))
        pReverse[idxMinus] <- rep(max(extendUpstream), length(idxMinus))

        pReverseMin <- rep(min(extendDownstream), length(pminGene))
        pReverseMin[idxMinus] <- rep(min(extendUpstream), length(idxMinus))

        pForward <- rep(max(extendUpstream), length(pminGene))
        pForward[idxMinus] <- rep(max(extendDownstream), length(idxMinus))      

        pForwardMin <- rep(min(extendUpstream), length(pminGene))
        pForwardMin[idxMinus] <- rep(min(extendDownstream), length(idxMinus))      

        ################################################################
        #We will test when genes pass by another gene promoter
        ################################################################

        #Start of Range is based on the max observed gene ranged <- direction
        s <- pmax(
          c(1, pmaxGene[-length(pmaxGene)] + tileSize), 
          pminGene - pReverse
        )
        s <- pmin(pminGene - pReverseMin, s)

        #End of Range is based on the max observed gene ranged -> direction
        e <- pmin(
            c(pminGene[-1] - tileSize, pmaxGene[length(pmaxGene)] + pForward[length(pmaxGene)]), 
            pmaxGene + pForward
          )
        e <- pmax(pmaxGene + pForwardMin, e)

        extendedGeneRegion <- IRanges(start = s, end = e)

        idx1 <- which(pminGene - pReverseMin < start(extendedGeneRegion))
        if(length(idx1) > 0){
          stop("Error in gene boundaries minError")
        }

        idx2 <- which(pmaxGene + pForwardMin > end(extendedGeneRegion))
        if(length(idx2) > 0){
          stop("Error in gene boundaries maxError")
        }
       
       rm(s, e, pReverse, pReverseMin, pForward, pForwardMin, geneStartz, geneEndz, pminGene, pmaxGene)

      }else{

        extendedGeneRegion <- ranges(suppressWarnings(extendGR(geneRegionz, upstream = max(extendUpstream), downstream = max(extendDownstream))))

      }

      tmp <- suppressWarnings(findOverlaps(extendedGeneRegion, uniqueTiles))
      x <- distance(ranges(geneRegionz)[queryHits(tmp)], uniqueTiles[subjectHits(tmp)])

      #Determine Sign for Distance relative to strand (Directionality determined based on dist from gene start)
      isMinus <- BiocGenerics::which(strand(geneRegionz) == "-")
      signDist <- sign(start(uniqueTiles)[subjectHits(tmp)] - start(resize(geneRegionz,1,"start"))[queryHits(tmp)])
      signDist[isMinus] <- signDist[isMinus] * -1

      #Correct the orientation for the distance!
      x <- x * signDist

      #Evaluate Input Model
      x <- eval(parse(text=geneModel))

      #Get Gene Weights Related to Gene Width
      x <- x * mcols(geneRegionz)$geneWeight[queryHits(tmp)]

      #Remove Blacklisted Tiles!
      if(!is.null(blacklist)){
        if(length(blacklist) > 0){
          blacklistz <- blacklist[[chrz]]
          if(is.null(blacklistz) | length(blacklistz) > 0){
            tilesBlacklist <- 1 * (!overlapsAny(uniqueTiles, ranges(blacklistz)))
            if(sum(tilesBlacklist == 0) > 0){
              x <- x * tilesBlacklist[subjectHits(tmp)] #Multiply Such That All Blacklisted Tiles weight is now 0!
            }
          }
        }
      }

      #Creating Sparse Matrix
      tmp <- Matrix::sparseMatrix(
        i = queryHits(tmp), 
        j = subjectHits(tmp), 
        x = x, 
        dims = c(length(geneRegionz), nrow(matGS))
      )

      #Calculate Gene Scores
      matGS <- tmp %*% matGS
      colnames(matGS) <- cellNames

      totalGSz <- Matrix::colSums(matGS)

      #Save tmp file
      ArchR:::.safeSaveRDS(matGS, file = paste0(tmpFile, "-", chrz, ".rds"), compress = FALSE)

      #Clean Memory
      rm(isMinus, signDist, extendedGeneRegion, uniqueTiles)
      rm(matGS, tmp)
      gc()

      totalGSz
   
    }, error = function(e){

      errorList <- list(
        ArrowFile = ArrowFile,
        geneRegions = geneRegions,
        blacklist = blacklist,
        chr = chrz,
        totalGSz = if(exists("totalGSz", inherits = FALSE)) totalGSz else "totalGSz",
        matGS = if(exists("matGS", inherits = FALSE)) matGS else "matGS"
      )

      ArchR:::.logError(e, fn = ".addGeneScoreMat TmpGS", info = sampleName, errorList = errorList, logFile = logFile)

    })

    totalGSz

  }, threads = subThreads) %>% Reduce("+", .)
  
  

  #########################################################################################################
  #Organize info for ArchR Arrow
  #########################################################################################################
  featureDF <- Reduce("c",geneRegions) %>% 
    {data.frame(
      row.names=NULL,
      seqnames=as.character(seqnames(.)),
      start=mcols(.)$geneStart,
      end=mcols(.)$geneEnd,
      strand=as.integer(strand(.)),
      name=mcols(.)$symbol,
      idx=mcols(.)$idx,
      stringsAsFactors=FALSE)}
  ArchR:::.logThis(featureDF, paste0(sampleName, " .addGeneScoreMat FeatureDF"), logFile = logFile)

  dfParams <- data.frame(
      extendUpstream = extendUpstream,
      extendDownstream = extendDownstream,
      geneUpstream = extendUpstream,
      geneDownstream = extendDownstream,
      scaleTo = scaleTo,
      tileSize = tileSize,
      ceiling = ceiling,
      geneModel = geneModel,
      stringsAsFactors=FALSE
    )

  ######################################
  # Initialize SP Mat Group
  ######################################
  o <- ArchR:::.initializeMat(
    ArrowFile = ArrowFile,
    Group = matrixName,
    Class = "double",
    Units = "NormCounts",
    cellNames = cellNames,
    params = dfParams,
    featureDF = featureDF,
    force = TRUE
  )

  #Clean Memory
  rm(dfParams, featureDF, genes)
  gc()

  #Normalize and add to Arrow File!
  for(z in seq_along(geneRegions)){

    o <- tryCatch({

      #Get Chromosome
      chrz <- paste0(unique(seqnames(geneRegions[[z]])))

      ArchR:::.logDiffTime(sprintf("Adding GeneScoreMatrix to %s for Chr (%s of %s)!", sampleName, z, length(geneRegions)), 
        tstart, verbose = FALSE, logFile = logFile)

      #Re-Create Matrix for that chromosome!
      matGS <- readRDS(paste0(tmpFile, "-", chrz, ".rds"))
      file.remove(paste0(tmpFile, "-", chrz, ".rds"))

      #Normalize
      # 2021.06.04 fumi
#      matGS@x <- as.numeric(scaleTo * matGS@x/rep.int(totalGS, Matrix::diff(matGS@p)))

      #Round to Reduce Digits After Final Normalization
      matGS@x <- round(matGS@x, 3)
      matGS <- Matrix::drop0(matGS)

      #Write sparseMatrix to Arrow File!
      o <- ArchR:::.addMatToArrow(
        mat = matGS, 
        ArrowFile = ArrowFile, 
        Group = paste0(matrixName, "/", chrz), 
        binarize = FALSE,
        addColSums = TRUE,
        addRowSums = TRUE,
        addRowVarsLog2 = TRUE #add for integration analyses
      )

      #Clean Memory
      rm(matGS)

      if(z %% 3 == 0 | z == length(geneRegions)){
        gc()
      }

    }, error = function(e){

      errorList <- list(
        ArrowFile = ArrowFile,
        geneRegions = geneRegions,
        blacklist = blacklist,
        chr = chrz,
        mat = if(exists("mat", inherits = FALSE)) mat else "mat"
      )

      ArchR:::.logError(e, fn = ".addGeneScoreMat AddToArrow", info = sampleName, errorList = errorList, logFile = logFile)

    })

  }

  return(ArrowFile)

}
```

```{r}
assignInNamespace(".addGeneScoreMat",
                  .myaddGeneScoreMat,
                  ns="ArchR")
```

# Add GeneScoreMatrix2

```{r}
proj6 = loadArchRProject("Save-Proj6/")
```

```{r}
addGeneScoreMatrix(
  proj6,
  matrixName = "GeneScoreMatrix2",
  geneScaleFactor = 1,
  force = TRUE)
```

```{r}
mat = getMatrixFromProject(proj6, useMatrix = "GeneScoreMatrix2")
assay(mat) = round(assay(mat))
```

# Library size normalization, as in sc-RNAseq

```{r}
library(Linnorm)
mat2 = expm1(Linnorm(assay(mat)))
# saveRDS(mat2, file = "mat2.rds")
```

Check

```{r}
max(assay(mat))
max(mat2)
summary(colSums(assay(mat)))
summary(colSums(mat2))
plot(assay(mat)[, 1], mat2[, 1])
```

# Impute GeneScoreMatrix, as in sc-RNAseq; save result

Run SAVER on different machines.
https://mohuangx.github.io/SAVER/articles/saver-tutorial.html

```{r}
library(SAVER)

mat2 = readRDS("mat2.rds")

# saver1 = saver(pred.genes = 1:3000, pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver1, file = "saver1.rds")
# 
# saver2 = saver(pred.genes = 3001:6000, pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver2, file = "saver2.rds")
# 
# saver3 = saver(pred.genes = 6001:9000, pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver3, file = "saver3.rds")
# 
# saver4 = saver(pred.genes = 9001:12000, pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver4, file = "saver4.rds")
# 
# saver5 = saver(pred.genes = 12001:15000, pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver5, file = "saver5.rds")
# 
# saver6 = saver(pred.genes = 15001:nrow(mat2), pred.genes.only = TRUE, do.fast = FALSE, x = mat2, ncores = 12, size.factor = 1)
# saveRDS(saver6, file = "saver6.rds")
# 
# saver1 = readRDS("saver1.rds")
# saver2 = readRDS("saver2.rds")
# saver3 = readRDS("saver3.rds")
# saver4 = readRDS("saver4.rds")
# saver5 = readRDS("saver5.rds")
# saver6 = readRDS("saver6.rds")
# saver.all <- combine.saver(list(saver1, saver2, saver3, saver4, saver5, saver6))
# saveRDS(saver.all,
#         file = "GeneScoreMatrix.geneScaleFactor_1.scaleTo_skipped.Linnorm.saver.rds")
# rm(saver1, saver2, saver3, saver4, saver5, saver6, saver.all)

saver.all = readRDS("GeneScoreMatrix.geneScaleFactor_1.scaleTo_skipped.Linnorm.saver.rds")
matsaver = saver.all$estimate
```

Check

```{r}
max(mat2)
max(matsaver)
min(mat2)
min(matsaver)
summary(colSums(mat2))
summary(colSums(matsaver))
plot(mat2[, 1], matsaver[, 1])
summary(rowMeans(mat2 == 0))
summary(rowMeans(matsaver == 0.001))

qplot(
  x = log10(rowMeans(matsaver)),
  # y = log10(rowSds(matsaver)),
  y = log10(rowSds(matsaver) / rowMeans(matsaver)), # CV
  geom = "bin2d")
```

# Check GeneScoreMatrix by SVD

```{r}
mat = getMatrixFromProject(proj6, useMatrix = "GeneScoreMatrix2")
colSm = colSums(assay(mat))

rownames(matsaver) = NULL
assay(mat) = log2(matsaver) # NOW LOG2 SCALE !!!
```

Simple cutoff of low expression genes

```{r}
quantile(rowMeans(assay(mat)), seq(0, 1, 0.1))
idx = (rowMeans(assay(mat)) >
         quantile(rowMeans(assay(mat)), 0.1)) 
# 0.1 0.2 0.3 0.5 0.7 similar for major coarse; 0.05 worse; maybe use 0.1
mat2 = mat[idx, ]
```

Library size correct

```{r}
assay(mat2) = assay(mat2) -
  matrix(
    colMeans(assay(mat2)),
    nrow = nrow(assay(mat2)),
    ncol = ncol(assay(mat2)),
    byrow = TRUE)
```

Sample batch correct mat2

```{r}
sampleLabels = unique(as.character(mat2$Sample))
ClusterscoarseLabels = unique(as.character(mat2$Clusterscoarse2))

for (cl in ClusterscoarseLabels) {
  print(cl)
  target = rowMeans(
    assay(mat2)[, which(mat2$Clusterscoarse2 == cl),
                drop = FALSE])
  for (s in sampleLabels) {
    print(s)
    x = assay(mat2)[, which(mat2$Clusterscoarse2 == cl &
                              mat2$Sample == s),
                    drop = FALSE]
    x = x - rowMeans(x) + target
    assay(mat2)[, which(mat2$Clusterscoarse2 == cl &
                          mat2$Sample == s)] = x
    rm(x)
  }
}
```

Row centralization is mandatory!
[Skillicorn, Understanding complex datasets. p.51]

```{r}
assay(mat2) = assay(mat2) -
  rowMeans(assay(mat2))
```

```{r}
gc()
set.seed(1)

svd <- irlba::irlba(assay(mat2), 30, 30)
```

Plot inspection

```{r}
library(viridisLite)

plotdata = as.data.frame(svd$v)
plotdata$colSm    = colSm
plotdata$celltype = mat2$Clusterscoarse2
plotdata$ploidy   = mat2$ploidy
plotdata$Sample   = as.character(mat2$Sample)

ggplot(
  data = plotdata,
  aes(x = V1,
      y = V2)) +
  geom_point(aes(
    # col = log10(colSm)
    col = celltype
    # col = ploidy
    ),
    size = 0.1) +
  # scale_color_gradientn(colours = viridis(256, option = "D")) +
  facet_grid(rows = vars(Sample)) +
  geom_rug()
```

# QC

```{r}
mat = getMatrixFromProject(proj6, useMatrix = "GeneScoreMatrix2")
rownames(matsaver) = NULL
assay(mat) = matsaver
```

The smallest value imputed by saver is 0.001
Exclude low expression genes.

```{r}
idx = which(rowMeans(assay(mat) <= 0.001) < 0.01) # discard 159, keep 17280
mat = mat[idx, ]
```

```{r}
assay(mat) = log2(assay(mat)) # NOW LOG2 SCALE !!!
```

```{r}
qplot(
  x = rowMeans(assay(mat)),
  y = rowSds(assay(mat)),
  geom = "bin2d")
```

```{r}
idx = which(rowMeans(assay(mat)) > log2(0.005)) # discard 2, keep 17278
mat = mat[idx, ]
```

```{r}
qplot(
  x = rowMeans(assay(mat)),
  y = rowSds(assay(mat)),
  geom = "bin2d")
```

Define rownames. Omit duplicates of the same gene

```{r}
dimnames(mat)[[1]] = rowData(mat)$name

idx = which(! duplicated(dimnames(mat)[[1]]))
mat = mat[idx, ]
```

Library size correct (qn)

```{r}
x = preprocessCore::normalize.quantiles(assay(mat))
assay(mat, withDimnames = FALSE) = x
```

This part is sidetrack.  Plot expression of cluster marker genes.

```{r}
markerGenes = c(
  "Alb", "Cyp2e1",
  "Krt19",
  "Stab2",
  "Des", "Lrat",
  "Csf3r",
  "Elane", "Ly6g6d", 
  "Csf1r",
  "Lyz2",
  "Marco",
  "Gata3",
  "Gzma",
  "Ebf1"
)

idx = which(rownames(mat) %in% markerGenes)
dataplot = as.data.frame(t(assay(mat[idx, ])))

g = proj6$Clusters4res06
g[g %in% paste0("C", 1:7)] = "hepatocyte"
g[g %in% paste0("C", 8:9)] = "stellate"
g[g %in% paste0("C", 11:13)] = "endothelial"
dataplot$Clusters4WBC = g[match(rownames(dataplot), proj6$cellNames)]

dataplot = tidyr::pivot_longer(
  dataplot,
  cols = -Clusters4WBC,
  names_to = "genename",
  values_to = "expression")
dataplot$Clusters4WBC = factor(
  dataplot$Clusters4WBC,
  levels = c("hepatocyte", "endothelial", "stellate", "C16", "C10", "C14", "C15"))
dataplot$genename = factor(
  dataplot$genename,
  levels = markerGenes)
dataplot$expression = 2^dataplot$expression

ggplot(
  data = dataplot,
  aes(x = Clusters4WBC,
      y = expression)) +
  geom_violin() +
  scale_y_log10() +
  facet_grid(rows = vars(genename), scales = "free_y") +
  labs(x = NULL,
       y = "Gene expression")
```

Row centralize

```{r}
assay(mat) = assay(mat) - rowMeans(assay(mat))
```

Compute and plot SVD without sample batch correction

```{r}
set.seed(1)
svd = irlba::irlba(assay(mat), nv = 20)
plot(svd$d)
ggplot(
  data = as.data.frame(svd$v)) +
  geom_point(aes(x = V1, y = V2), size = 0.3) +
  facet_grid(
    cols = vars(mat$Clusters4BTmyelo),
    rows = vars(as.character(mat$Sample)))
```

The first two components capture
A. sample batch effect, and
B. difference between major cell types.
In the following, we analyze the effect of diet intervention by comparing samples.
Specifically, we compare nuclei of same cell type between different sample.
Thus, we prefer to eliminate A, hoping that the removal of B does not harm.

```{r}
assay(mat) = assay(mat) -
  svd$u[, 1:2] %*% diag(svd$d[1:2]) %*% t(svd$v[, 1:2])
```

# Save result

```{r}
saveRDS(mat, file = "matGeneScoreSamplebatchcorrected.rds")
```

# Compare between conditions

```{r}
proj6 = loadArchRProject("Save-Proj6/")
mat = readRDS("matGeneScoreSamplebatchcorrected.rds")
```

```{r}
g = proj6$Sample
g[g %in% c("m154211", "m167108")] = "sp"
proj6$spdietcombined = g

proj6$Clusters4BTmyelospdietcombined =
  paste0(proj6$Clusters4BTmyelo,
         "_",
         proj6$spdietcombined)
```

```{r}
gname = "Clusters4BTmyelospdietcombined"
g = getCellColData(proj6, gname, drop = TRUE)[
  match(colnames(mat),
        proj6$cellNames)]

resultcompared =
  do.call(
    rbind,
    lapply(
      c("hepato", "stellate", "endothelial", "myelo", "B", "T_NK"),
      function (celltype) {
        do.call(
          rbind,
          lapply(
            c("m154207", "m167203", "m168101"),
            function (s) {
              tgt = paste0(celltype, "_", s)
              bgd = paste0(celltype, "_sp")
              print(c(tgt, bgd))
              
              cond = rep(NA, ncol(mat))
              cond[g == tgt] = 1
              cond[g == bgd] = 0
              a0 = lm(y ~ x,
                      data = list(
                        y = t(assay(mat)),
                        x = cond))
              resulttgtbgd = broom::tidy(a0)
              resulttgtbgd = resulttgtbgd[resulttgtbgd$term == "x", ]
              colnames(resulttgtbgd)[colnames(resulttgtbgd) == "p.value"] = "p"
              resulttgtbgd$logp = pt(- abs(resulttgtbgd$statistic),
                                     df = a0$df.residual,
                                     log.p = TRUE) +
                log(2)
              resulttgtbgd$FDR = p.adjust(resulttgtbgd$p, method = "BH")
              colnames(resulttgtbgd)[colnames(resulttgtbgd) == "response"] = "genename"
              resulttgtbgd$ENTREZID =
                AnnotationDbi::mapIds(
                  org.Rn.eg.db,
                  keys = resulttgtbgd$genename,
                  column = "ENTREZID",
                  keytype = "SYMBOL",
                  multiVals = "first")
              resulttgtbgd$celltype = celltype
              resulttgtbgd$sample = s
              resulttgtbgd$tgt = tgt
              resulttgtbgd$bgd = bgd
              return(resulttgtbgd)
            }))
      }))
resultcompared$celltype =
  factor(resultcompared$celltype,
         levels = c("hepato", "stellate", "endothelial", "myelo", "B", "T_NK"))
```

Show top hits

```{r}
data =
resultcompared %>%
  dplyr::mutate(foldchange = 2^estimate) %>%
  dplyr::select(-c("term", "estimate", "std.error", "FDR", "tgt", "bgd")) %>%
  dplyr::group_by(celltype, sample) %>%
  dplyr::mutate(rank = rank(statistic)) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(celltype) %>%
  tidyr::pivot_wider(names_from = sample, values_from = c("foldchange", "statistic", "p", "logp", "rank")) %>%
  dplyr::filter(pmin(foldchange_m154207, foldchange_m167203, foldchange_m168101) < 1/1.1 |
                  pmax(foldchange_m154207, foldchange_m167203, foldchange_m168101) > 1.1) %>%
  dplyr::mutate(pminrank = pmin(rank_m154207, rank_m167203, rank_m168101),
                pmaxrank = pmax(rank_m154207, rank_m167203, rank_m168101)) %>%
  dplyr::filter(pminrank <= 3 | pmaxrank > max(pmaxrank) - 3) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(- (statistic_m154207 + statistic_m167203 + statistic_m168101)) %>%
  dplyr::arrange(celltype)
data$chr = as.character(seqnames(geneAnnotation$genes))[
  match(data$genename,
        geneAnnotation$genes@elementMetadata$gene_id)]
data$start = start(geneAnnotation$genes)[
  match(data$genename,
        geneAnnotation$genes@elementMetadata$gene_id)]
```

# Cross-check with bulk differential gene expression experiments

```{r}
# HFD 4w
s = "m154207"
foo = read.csv("~/human/rat_mRNA_liver/Liver_diet.ComBat.csv", header = TRUE)
foo$Z = foo$log2_foldchange_dietHFC / foo$SE_dietHFC

# washout
s = "m168101"
foo = read.csv("~/human/rat_mRNA_liver/Liver_HFC20w.csv", header = TRUE)
foo$Z = foo$log2_foldchange_HFC / foo$SE_HFC

foo = foo[, c("GeneName", "Z")]
foo = foo[(foo$GeneName != "") & ! is.na(foo$Z), ]
dataplot = resultcompared %>%
  dplyr::filter(sample == s)
dataplot$bulkZ = foo$Z[match(dataplot$genename, foo$GeneName)]
dataplot$width = width(geneAnnotation$genes)[
  match(dataplot$genename,
        geneAnnotation$genes@elementMetadata$gene_id)]
dataplot$widthclass = "short"
dataplot$widthclass[dataplot$width > median(dataplot$width)] = "long"

dataplot = dataplot[
  dataplot$celltype %in%
    c("hepato", "stellate", "endothelial", "myelo"), ]
dataplot$celltype =
  factor(as.character(dataplot$celltype),
         levels = c("hepato", "endothelial", "stellate", "myelo"),
         labels = c("Hepatocyte", "Endothelial", "Stellate", "Macrophage"))

p1 =
  ggplot(data = dataplot,
         aes(x = statistic,
             y = bulkZ)) +
  geom_point(size = 0.5) +
  stat_smooth() +
  facet_grid(
    # rows = vars(widthclass),
    cols = vars(celltype),
    scales = "free_x") +
  labs(title = s,
       x = "Z for DGE in snATAC-seq",
       y = "Z for DGE in bulk mRNA")
ggsave(p1, filename = "crosscheckbulk.tiff", width = 9, height = 3, dpi = 150)
```

# Unify highly correlated genes in mat

Target cell type

```{r}
tgtct =
  "hepato"
  # "stellate"
  # "endothelial"
  # "myelo"
  # "T_NK"
```

Since highly correlated genes could be problematic in downstream analysis,
unify those to a representative.

```{r}
mattgtct = mat[, which(mat$Clusters4BTmyelo == tgtct)]

matcor = cor(t(assay(mattgtct)), method = "spearman") # takes an hour

diag(matcor) = 0
quantile(rowMaxs(abs(matcor)), seq(0, 1, 0.01))
quantile(rowMaxs(abs(matcor)), seq(0.99, 1, 0.001))
matcor[abs(matcor) < 0.9] = 0

# Are there negative correlations?
min(matcor)
matcor = abs(matcor)

# omit unique genes/motifs
index = which(rowSums(matcor) > 0)
identical(index, which(colSums(matcor) > 0))
matcor = matcor[index, index]

library(igraph)
x = as.character()
for (j in 1:ncol(matcor)) {
  connected = matcor[, j]
  connected = connected[connected > 0]
  x = c(x,
        as.character(
          rbind(
            names(connected),
            colnames(matcor)[j])))
}
matcorgraph = graph(edges = x, directed = FALSE)
matcorgraphcomp = igraph::components(matcorgraph, mode = "weak")

redundant = list()
for (i in unique(matcorgraphcomp$membership)) {
  m = names(matcorgraphcomp$membership)[matcorgraphcomp$membership == i]
  m = names(sort(rowSums(matcor[m, ]), decreasing = TRUE))
  # the largest is kept (and stored as name), and the others are redundant
  x = list(m[-1])
  names(x) = m[1]
  redundant = c(redundant, x)
}

# non-redundant
index = which(! rownames(mattgtct) %in% unlist(redundant))
matnr = mattgtct[index, ]

redundantGene = redundant
matnrGene = matnr
rownames(matnrGene) = paste0("Gene.", rownames(matnrGene))

# saveRDS(redundantGene,  file = paste0("redundantGene.",  tgtct, ".rds"))
# saveRDS(matnrGene,      file = paste0("matnrGene.",      tgtct, ".rds"))
redundantGene  = readRDS(paste0("redundantGene.",  tgtct, ".rds"))
matnrGene      = readRDS(paste0("matnrGene.",      tgtct, ".rds"))
```
