---
title: "MotifMatrix"
output: html_notebook
---

# Add MotifMatrix

```{r}
proj6 = loadArchRProject("Save-Proj6/")
```

The following computation took 16 mins

```{r}
proj6 <- addMotifAnnotations(
  ArchRProj = proj6,
  motifSet = "cisbp",
  species = "Mus musculus",
  name = "Motif")

proj6@peakAnnotation@listData$Motif$motifSummary
```

The following computation took 54 mins

```{r}
proj6 <- addBgdPeaks(proj6)

proj6 <- addDeviationsMatrix(
  ArchRProj = proj6, 
  peakAnnotation = "Motif",
  force = TRUE)
```

```{r}
mat = getMatrixFromProject(proj6, "MotifMatrix")
mat@assays@data@listData =
  mat@assays@data@listData["z"] # drop deviations assay

idx = which(rowSums(is.na(assay(mat))) == 0)
mat = mat[idx, ]
```

Convert rownames to rat gene symbol

```{r}
dimnames(mat)[[1]] = 
  sub("_.*", "", dimnames(mat)[[1]])
# Tcfap2a_1 -> Tcfap2a

attach("~/human/publicrat/homologs/mouse_to_rat_homologs.Rda")
x = mouse_to_rat_homologs$RGD.symbol[
  match(
    rownames(mat),
    mouse_to_rat_homologs$NCBI.gene..formerly.Entrezgene..accession)]
dimnames(mat)[[1]] = x
idx = which((! is.na(rownames(mat))) & (rownames(mat) != ""))
mat = mat[idx, ]

idx = which(! duplicated(rownames(mat)))
mat = mat[idx, ]
```

# QC

Library size correct (qn)

```{r}
assay(mat) = as.matrix(assay(mat))
x = preprocessCore::normalize.quantiles(assay(mat))
assay(mat, withDimnames = FALSE) = x
```

Row centralize

```{r}
assay(mat) = assay(mat) - rowMeans(assay(mat))
```

Compute and plot SVD without sample batch correction

```{r}
set.seed(1)
svd = irlba::irlba(assay(mat), nv = 20)
plot(svd$d)
ggplot(
  data = as.data.frame(svd$v)) +
  geom_point(aes(x = V1, y = V2), size = 0.3) +
  facet_grid(
    cols = vars(mat$Clusters4BTmyelo),
    rows = vars(as.character(mat$Sample)))
```

The first two components capture
A. sample batch effect, and
B. difference between major cell types.
In the following, we analyze the effect of diet intervention by comparing samples.
Specifically, we compare nuclei of same cell type between different sample.
Thus, we prefer to eliminate A, hoping that the removal of B does not harm.

```{r}
assay(mat) = assay(mat) -
  svd$u[, 1:2] %*% diag(svd$d[1:2]) %*% t(svd$v[, 1:2])
```

# Save result

```{r}
saveRDS(mat, file = "matMotifSamplebatchcorrected.rds")
```

# Compare between conditions

```{r}
proj6 = loadArchRProject("Save-Proj6/")
mat = readRDS("matMotifSamplebatchcorrected.rds")
```

```{r}
g = proj6$Sample
g[g %in% c("m154211", "m167108")] = "sp"
proj6$spdietcombined = g

proj6$Clusters4BTmyelospdietcombined =
  paste0(proj6$Clusters4BTmyelo,
         "_",
         proj6$spdietcombined)
```

```{r}
gname = "Clusters4BTmyelospdietcombined"
g = getCellColData(proj6, gname, drop = TRUE)[
  match(colnames(mat),
        proj6$cellNames)]

resultcompared =
  do.call(
    rbind,
    lapply(
      c("hepato", "stellate", "endothelial", "myelo", "B", "T_NK"),
      function (celltype) {
        do.call(
          rbind,
          lapply(
            c("m154207", "m167203", "m168101"),
            function (s) {
              tgt = paste0(celltype, "_", s)
              bgd = paste0(celltype, "_sp")
              print(c(tgt, bgd))
              
              cond = rep(NA, ncol(mat))
              cond[g == tgt] = 1
              cond[g == bgd] = 0
              resulttgtbgd = broom::tidy(
                lm(y ~ x,
                   data = list(
                     y = t(assay(mat)),
                     x = cond)))
              resulttgtbgd = resulttgtbgd[resulttgtbgd$term == "x", ]
              colnames(resulttgtbgd)[colnames(resulttgtbgd) == "p.value"] = "p"
              resulttgtbgd$FDR = p.adjust(resulttgtbgd$p, method = "BH")
              colnames(resulttgtbgd)[colnames(resulttgtbgd) == "response"] = "genename"
              resulttgtbgd$ENTREZID =
                AnnotationDbi::mapIds(
                  org.Rn.eg.db,
                  keys = resulttgtbgd$genename,
                  column = "ENTREZID",
                  keytype = "SYMBOL",
                  multiVals = "first")
              resulttgtbgd$celltype = celltype
              resulttgtbgd$sample = s
              resulttgtbgd$tgt = tgt
              resulttgtbgd$bgd = bgd
              return(resulttgtbgd)
            }))
      }))
resultcompared$celltype =
  factor(resultcompared$celltype,
         levels = c("hepato", "stellate", "endothelial", "myelo", "B", "T_NK"))
```

# Unify highly correlated motifs in mat

Target cell type

```{r}
tgtct =
  "hepato"
  # "stellate"
  # "endothelial"
  # "myelo"
  # "T_NK"
```

Since highly correlated motifs could be problematic in downstream analysis,
unify those to a representative.

```{r}
mattgtct = mat[, which(mat$Clusters4BTmyelo == tgtct)]

matcor = cor(t(assay(mattgtct)), method = "spearman")

diag(matcor) = 0
quantile(rowMaxs(abs(matcor)), seq(0, 1, 0.01))
quantile(rowMaxs(abs(matcor)), seq(0.99, 1, 0.001))
matcor[abs(matcor) < 0.9] = 0

# Are there negative correlations?
min(matcor)
matcor = abs(matcor)

# omit unique genes/motifs
index = which(rowSums(matcor) > 0)
identical(index, which(colSums(matcor) > 0))
matcor = matcor[index, index]

library(igraph)
x = as.character()
for (j in 1:ncol(matcor)) {
  connected = matcor[, j]
  connected = connected[connected > 0]
  x = c(x,
        as.character(
          rbind(
            names(connected),
            colnames(matcor)[j])))
}
matcorgraph = graph(edges = x, directed = FALSE)
matcorgraphcomp = igraph::components(matcorgraph, mode = "weak")

redundant = list()
for (i in unique(matcorgraphcomp$membership)) {
  m = names(matcorgraphcomp$membership)[matcorgraphcomp$membership == i]
  m = names(sort(rowSums(matcor[m, ]), decreasing = TRUE))
  # the largest is kept (and stored as name), and the others are redundant
  x = list(m[-1])
  names(x) = m[1]
  redundant = c(redundant, x)
}

# non-redundant
index = which(! rownames(mattgtct) %in% unlist(redundant))
matnr = mattgtct[index, ]

redundantMotif = redundant
matnrMotif = matnr
rownames(matnrMotif) = paste0("Motif.", rownames(matnrMotif))

# saveRDS(redundantMotif, file = paste0("redundantMotif.", tgtct, ".rds"))
# saveRDS(matnrMotif,     file = paste0("matnrMotif.",     tgtct, ".rds"))
redundantMotif = readRDS(paste0("redundantMotif.", tgtct, ".rds"))
matnrMotif     = readRDS(paste0("matnrMotif.",     tgtct, ".rds"))
```
