---
title: "GSEA"
output: html_notebook
---

# Load GeneScoreMatrix

```{r}
proj6 = loadArchRProject("Save-Proj6/")
mat = readRDS("matGeneScoreSamplebatchcorrected.rds")
```

# Define gene sets for GSEA

```{r}
library(msigdbr)

gslistfrombaderlab = function (x) {
  x = strsplit(x, "\t")
  gslist = lapply(x, function (x) { x[-c(1:2)] })
  names(gslist) = unlist(lapply(x, function (x) { x[1] }))
  return(gslist)
}
gslistfrommsigdbr = function (x) {
  x = x %>%
    dplyr::mutate(name = paste0(gs_name, "%%", gs_id)) %>%
    dplyr::select(name, gene_symbol) %>%
    dplyr::group_by(name) %>%
    dplyr::summarize(gs = list(gene_symbol))
  gslist = x$gs
  names(gslist) = x$name
  return(gslist)
}

gslists = list()
x = readLines("~/human/publicrat/baderlab_EM_Genesets/Rat_Human_NetPath_June_01_2021_symbol.gmt")
gslists = c(gslists,
            list(NetPath = gslistfrombaderlab(x)))
x = readLines("~/human/publicrat/baderlab_EM_Genesets/Rat_Human_Panther_June_01_2021_symbol.gmt")
gslists = c(gslists,
            list(Panther = gslistfrombaderlab(x)))
x = msigdbr(species = "Rattus norvegicus", category = "H")
gslists = c(gslists,
            list(H = gslistfrommsigdbr(x)))
x = msigdbr(species = "Rattus norvegicus", category = "C2", subcategory = "CP:KEGG")
gslists = c(gslists,
            list(KEGG = gslistfrommsigdbr(x)))
x = msigdbr(species = "Rattus norvegicus", category = "C2", subcategory = "CP:PID")
gslists = c(gslists,
            list(PID = gslistfrommsigdbr(x)))
x = msigdbr(species = "Rattus norvegicus", category = "C2", subcategory = "CP:REACTOME")
gslists = c(gslists,
            list(REACTOME = gslistfrommsigdbr(x)))
x = msigdbr(species = "Rattus norvegicus", category = "C2", subcategory = "CP:WIKIPATHWAYS")
gslists = c(gslists,
            list(WIKIPATHWAYS = gslistfrommsigdbr(x)))
x1 = msigdbr(species = "Rattus norvegicus", category = "C3", subcategory = "TFT:GTRD")
x2 = msigdbr(species = "Rattus norvegicus", category = "C3", subcategory = "TFT:TFT_Legacy")
gslists = c(gslists,
            list(TFT = c(gslistfrommsigdbr(x1), gslistfrommsigdbr(x2))))
x = msigdbr(species = "Rattus norvegicus", category = "C5", subcategory = "GO:BP")
gslists = c(gslists,
            list(GOBP = gslistfrommsigdbr(x)))
rm(x, x1, x2)
```

# Define conditions to be compared

```{r}
g = proj6$Sample
g[g %in% c("m154211", "m167108")] = "sp"
proj6$spdietcombined = g

proj6$Clusters4BTmyelospdietcombined =
  paste0(proj6$Clusters4BTmyelo,
         "_",
         proj6$spdietcombined)
```

Compare nuclei of same cell type between different interventions
Choose tgt vs bgd

```{r}
gname = "Clusters4BTmyelospdietcombined"
g = getCellColData(proj6, gname, drop = TRUE)[
  match(colnames(mat),
        proj6$cellNames)]
ClustersLabels = unique(as.character(g))

tgt = "hepato_m154207"; bgd = "hepato_sp"
tgt = "hepato_m167203"; bgd = "hepato_sp"
tgt = "hepato_m168101"; bgd = "hepato_sp"

tgt = "stellate_m154207"; bgd = "stellate_sp"
tgt = "stellate_m167203"; bgd = "stellate_sp"
tgt = "stellate_m168101"; bgd = "stellate_sp"

tgt = "endothelial_m154207"; bgd = "endothelial_sp"
tgt = "endothelial_m167203"; bgd = "endothelial_sp"
tgt = "endothelial_m168101"; bgd = "endothelial_sp"

tgt = "myelo_m154207";  bgd = "myelo_sp"
tgt = "myelo_m167203";  bgd = "myelo_sp"
tgt = "myelo_m168101";  bgd = "myelo_sp"

tgt = "B_m154207";  bgd = "B_sp"
tgt = "B_m167203";  bgd = "B_sp"
tgt = "B_m168101";  bgd = "B_sp"

tgt = "T_NK_m154207";  bgd = "T_NK_sp"
tgt = "T_NK_m167203";  bgd = "T_NK_sp"
tgt = "T_NK_m168101";  bgd = "T_NK_sp"
```

Compare clusters

```{r}
gname = "Clusters4res06"
g = getCellColData(proj6, gname, drop = TRUE)[
  match(colnames(mat),
        proj6$cellNames)]
ClustersLabels = unique(as.character(g))
ClustersLabels = ClustersLabels[
  order(as.numeric(sub("^C", "", ClustersLabels)))]

tgt = "C16"; bgd = "C10" # inflammatory vs non-inflammatory macrophage
tgt =  "C8"; bgd =  "C9" # stellate
tgt = "C11"; bgd = "C12" # endothelial cell
```

# Run padog GSEA

```{r}
esetm = cbind(
  assay(mat)[, which(g == tgt)],
  assay(mat)[, which(g == bgd)])
group = c(rep("d", sum(g == tgt)), rep("c", sum(g == bgd)))

library(PADOG)
dseed = 1
result = list()
for (i in 1:length(gslists)) {
  print(names(gslists)[i])
  myr = padog(
    esetm = esetm,
    group = group,
    gslist = gslists[[i]],
    organism = "rno",
    Nmin = 5,
    NI = max(10000, length(gslists[[i]]) * 20),
    parallel = TRUE,
    ncr = 16,
    dseed = dseed)
  myr$FDRmeanAbsT = p.adjust(myr$PmeanAbsT, method = "BH")
  myr$FDRpadog    = p.adjust(myr$Ppadog, method = "BH")
  result = c(result, list(myr))
}
names(result) = names(gslists)

saveRDS(result, file = paste0(
  "padog/",
  paste0(c(gname, tgt, bgd, dseed, "rds"), collapse = ".")))
```

# Inspect enriched gene sets

When there are too many gene sets attaining FDRpadog < 0.05,
visualize and choose representatives by using Cytoscape:
1. Apps > EnrichmentMap. Import GMT file `foo.gmt`
2. Node Table. Import `foo.txt` and add column log10Ppadog
3. Style > Fill Color > Map. Column = log10Ppadog. Mapping Type = Continuous Mapping
4. [Optional] Apps > clusterMaker > MCODE Cluster. Create new clustered network
5. Layout > yFiles Radial Layout

As above, define `gname`, `tgt`, `bgd`.
Generate files to be loaded to Cytoscape.

```{r}
dseed = 1
padog =
  readRDS(paste0(
    "padog/",
    # "padog.GeneScore/",
    # "padog.Motif/",
    paste0(c(gname, tgt, bgd, dseed, "rds"), collapse = ".")))

x = padog[c(1:7, 9)]
x = do.call(rbind, x)
padogFDR005 = x[x$FDRpadog < 0.05, ]
padogFDR005$log10Ppadog = log10(padogFDR005$Ppadog)
write.table(
  padogFDR005[, c("Name", "log10Ppadog")],
  file = "foo.txt",
  row.names = FALSE,
  col.names = TRUE,
  sep = "\t",
  quote = FALSE)

output = do.call(c, gslists)
output = output[rownames(padogFDR005)]
output = lapply(output, function (x) { paste0(x, collapse = "\t") })
x = names(output)
x = sub("[^.]*\\.", "", x, perl = TRUE)
output = paste(
  x,
  sub("%.*", "", x),
  unlist(output),
  sep = "\t")
cat(output, file = "foo.gmt", sep = "\n")
```

Generate files to be loaded to Revigo.
Revigo clusters GO terms. It doesn't handle gene sets from other sources.
Uses `padog` defined in previous chunk.

```{r}
x = padog$GOBP
padogFDR005 = x[x$FDRpadog < 0.05, ]
padogFDR005$gs_id = sub(".*%%", "", padogFDR005$Name)
x = msigdbr(species = "Rattus norvegicus", category = "C5", subcategory = "GO:BP")
padogFDR005$gs_exact_source =
  x$gs_exact_source[
    match(padogFDR005$gs_id,
          x$gs_id)]
write.table(
  padogFDR005[, c("gs_exact_source", "Ppadog")],
  file = "foo.txt",
  row.names = FALSE,
  col.names = FALSE,
  sep = "\t",
  quote = FALSE)
```

# Summarize results across tgt/bgd combinations

For all comparisons under `gname = "Clusters4BTmyelospdietcombined"`,
summarize the results for a given geneset `gsname` in databaset `dbname`.

```{r}
do.call(
  rbind,
  lapply(
    c("hepato", "stellate", "endothelial", "myelo", "T_NK"),
    function (celltype) {
      do.call(
        rbind,
        lapply(
          c("m154207", "m167203", "m168101"),
          function (s) {
            tgt = paste0(celltype, "_", s)
            bgd = paste0(celltype, "_sp")
            x =
              readRDS(paste0("padog/", paste0(c(gname, tgt, bgd, dseed, "rds"), collapse = ".")))
            x = x[[dbname]][gsname, ]
            x$celltype = celltype
            x$sample = s
            x$score =
              3 * (x$FDRpadog <  0.05) +
              2 * (x$FDRpadog >= 0.05 & x$Ppadog <  0.01) +
              1 * (x$FDRpadog >= 0.05 & x$Ppadog >= 0.01 & x$Ppadog < 0.05)
            return(x)
          }))
    }))
```

Barcode plot
Beforehand, compute `resultcompared` in 7_GeneScoreMatrix.Rmd

```{r}
gs = gslists[[dbname]][[gsname]]
pdf("foo.pdf", width = 10, height = 10)
par(mfrow = c(5, 3), mai = rep(0.25, 4))
for (celltype in c("hepato", "stellate", "endothelial", "myelo", "T_NK")) {
  for (s in c("m154207", "m167203", "m168101")) {
    tgt = paste0(celltype, "_", s)
    bgd = paste0(celltype, "_sp")
    print(c(tgt, bgd))
    dataplot = resultcompared[
      (resultcompared$tgt == tgt &
         resultcompared$bgd == bgd), ]
    dataplot$geneset = dataplot$genename %in% gs
    limma::barcodeplot(
      dataplot$statistic,
      dataplot$geneset,
      labels = c(bgd, tgt),
      quantiles = c(-1,1) * qnorm(0.025, lower.tail = FALSE))
  }
}
dev.off()
```

# Inspection of individual up & down regulated genes in pathway

Uses `dbname`, `gsname`, `tgt`, `bgd`.

```{r}
gs = gslists[[dbname]][[gsname]]
output = resultcompared[
  (resultcompared$tgt == tgt &
     resultcompared$bgd == bgd), ]
output = output[output$genename %in% gs, ]
output = output[, c("genename", "ENTREZID", "statistic")]
write.csv(output, "foo.csv")
```

Cyotoscape cannot import KEGG properly. Use KEGG website.
KEGG Mapper – Convert ID; NCBI-GeneID -> KEGG GENES
https://www.kegg.jp/kegg/tool/conv_id.html
KEGG Mapper – Color Pathway
https://www.genome.jp/kegg/tool/map_pathway3.html

# Volcano plot

Uses `gs`.

```{r}
plotdata = resultcompared %>%
  filter(genename %in% gs)
plotdata$log10p = log10(plotdata$p)
x = is.infinite(plotdata$log10p)
plotdata$log10p[x] = min(plotdata$log10p[! x])

fig = ggplot(data = plotdata) +
  geom_point(
    aes(x = estimate,
        y = - log10p,
        col = FDR < 0.01),
    size = 0.3) +
  facet_grid(
    rows = vars(celltype),
    cols = vars(sample),
    scales = "free_y") +
  scale_color_manual(values = c("black", "red")) +
  labs(
    x = "log2(Fold change)",
    y = "-log10(P)")
ggsave("foo.pdf", fig, width = 6, height = 8)
```

# Heatmap plot of top genes

Uses `topgenes`, `topgenesbytargetdbgs`.

```{r}
plotdata = resultcompared %>%
  filter(genename %in% topgenes) %>%
  filter(celltype != "B")
plotdata$estimate[plotdata$FDR >= 0.01] = 0
plotdata = as.data.frame(
  tidyr::pivot_wider(
    plotdata[, c("genename", "estimate", "tgt")],
    names_from = tgt,
    values_from = estimate))
row.names(plotdata) = plotdata[, 1]
plotdata = plotdata[, -1]
plotdata = as.matrix(plotdata)
plotdata = plotdata[match(topgenes, row.names(plotdata)), ]

library(gplots)

pdf(file = "heatmap.pdf", width = 6, height = 8)
colfunc = colorRampPalette(c("blue", "white", "red"))
heatmap.2(
  plotdata,
  Colv = NA,
  Rowv = as.dendrogram(hclust(as.dist(1 - cor(t(plotdata))), method = "ward.D2")),
  dendrogram = "row",
  scale = "none",
  col = colfunc(256),
  colsep = seq(0, ncol(plotdata), 3),
  rowsep = c(0, nrow(plotdata)),
  sepcolor = "Gray",
  cexCol = 0.5,
  trace = "none",
  density.info = "none",
  key.title = NA,
  key.xlab = "log2(Fold change)")
dev.off()

colfunc = colorRampPalette(c("white", "black"))
heatmap.2(
  topgenesbytargetdbgs,
  Colv = NA,
  Rowv = as.dendrogram(hclust(as.dist(1 - cor(t(plotdata))), method = "ward.D2")),
  dendrogram = "none",
  scale = "none",
  col = colfunc(2),
  colsep = c(0, ncol(topgenesbytargetdbgs)),
  rowsep = c(0, nrow(topgenesbytargetdbgs)),
  sepcolor = "Gray",
  cexCol = 0.75,
  trace = "none",
  density.info = "none")
```

# Result: representative enriched gene sets

```{r}
gname = "Clusters4BTmyelospdietcombined"

tgt = "hepato_m154207"; bgd = "hepato_sp"
# dbname ="Panther"; gsname = "CHOLESTEROL BIOSYNTHESIS%PANTHER PATHWAY%P00014"
# dbname ="H"; gsname = "HALLMARK_XENOBIOTIC_METABOLISM%%M5934"
# dbname ="H"; gsname = "HALLMARK_BILE_ACID_METABOLISM%%M5948"
# dbname ="H"; gsname = "HALLMARK_CHOLESTEROL_HOMEOSTASIS%%M5892"
# dbname ="KEGG"; gsname = "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS%%M3214"
# dbname ="KEGG"; gsname = "KEGG_STEROID_BIOSYNTHESIS%%M5872"
# dbname ="PID"; gsname = "PID_HNF3B_PATHWAY%%M106"
# dbname ="PID"; gsname = "PID_IL2_PI3K_PATHWAY%%M143"
dbname ="REACTOME"; gsname = "REACTOME_METABOLISM_OF_STEROIDS%%M27832"
dbname ="REACTOME"; gsname = "REACTOME_CHOLINE_CATABOLISM%%M27619"
# dbname ="REACTOME"; gsname = "REACTOME_BILE_ACID_AND_BILE_SALT_METABOLISM%%M499"
# dbname ="REACTOME"; gsname = "REACTOME_XENOBIOTICS%%M5372"
# dbname ="REACTOME"; gsname = "REACTOME_CHOLESTEROL_BIOSYNTHESIS%%M16227"
# dbname ="REACTOME"; gsname = "REACTOME_FATTY_ACID_METABOLISM%%M27854"
# dbname ="WIKIPATHWAYS"; gsname = "WP_CHOLESTEROL_BIOSYNTHESIS_PATHWAY%%M39608"
dbname ="WIKIPATHWAYS"; gsname = "WP_SIGNAL_TRANSDUCTION_THROUGH_IL1R%%M39803"
dbname ="GOBP"; gsname = "GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS%%M10829"
dbname ="GOBP"; gsname = "GOBP_FATTY_ACID_METABOLIC_PROCESS%%M17829"
dbname ="GOBP"; gsname = "GOBP_LIPID_LOCALIZATION%%M15749"
# dbname ="GOBP"; gsname = "GOBP_LIPID_BIOSYNTHETIC_PROCESS%%M14504"
# dbname ="GOBP"; gsname = "GOBP_CHOLINE_CATABOLIC_PROCESS%%M23533"
# dbname ="GOBP"; gsname = "GOBP_STEROID_BIOSYNTHETIC_PROCESS%%M13381"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_DRUG%%M6886"
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_LIPID_LOCALIZATION%%M25391"
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_LIPID_TRANSPORT%%M13620"
# dbname ="GOBP"; gsname = "GOBP_BILE_ACID_BIOSYNTHETIC_PROCESS%%M12041"
tgt = "hepato_m167203"; bgd = "hepato_sp"
# dbname ="NetPath"; gsname = "IL1%NETPATH%IL1"  # up Il1r1 Il1r2 Nfkb1 relevant? NetPath reliable?
dbname ="Panther"; gsname = "APOPTOSIS SIGNALING PATHWAY%PANTHER PATHWAY%P00006"  # up
# dbname ="Panther"; gsname = "PLASMINOGEN ACTIVATING CASCADE%PANTHER PATHWAY%P00050"  # up
dbname ="H"; gsname = "HALLMARK_TNFA_SIGNALING_VIA_NFKB%%M5890"
# dbname ="H"; gsname = "HALLMARK_IL6_JAK_STAT3_SIGNALING%%M5897"
dbname ="KEGG"; gsname = "KEGG_APOPTOSIS%%M8492"   #!!! rn7
# dbname ="PID"; gsname = "PID_IL1_PATHWAY%%M110"
# dbname ="PID"; gsname = "PID_IL2_PI3K_PATHWAY%%M143"
dbname ="REACTOME"; gsname = "REACTOME_RESPONSE_TO_METAL_IONS%%M27541"
# dbname ="REACTOME"; gsname = "REACTOME_INTERLEUKIN_10_SIGNALING%%M27605"
dbname ="REACTOME"; gsname = "REACTOME_INTERLEUKIN_1_SIGNALING%%M27895"
# dbname ="REACTOME"; gsname = "REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM%%M1060"
# dbname ="REACTOME"; gsname = "REACTOME_SIGNALING_BY_INTERLEUKINS%%M874"
# dbname ="REACTOME"; gsname = "REACTOME_CYP2E1_REACTIONS%%M27130"
# dbname ="REACTOME"; gsname = "REACTOME_XENOBIOTICS%%M5372"
# dbname ="REACTOME"; gsname = "REACTOME_BILE_ACID_AND_BILE_SALT_METABOLISM%%M499"
dbname ="REACTOME"; gsname = "REACTOME_INFLAMMASOMES%%M1072"
# dbname ="WIKIPATHWAYS"; gsname = "WP_NUCLEAR_RECEPTORS_IN_LIPID_METABOLISM_AND_TOXICITY%%M39488"
dbname ="WIKIPATHWAYS"; gsname = "WP_SIGNAL_TRANSDUCTION_THROUGH_IL1R%%M39803"
# dbname ="TFT"; gsname = "NFKB_C%%M12240"
dbname ="GOBP"; gsname = "GOBP_CYTOKINE_PRODUCTION%%M13156"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_CELL_CELL_ADHESION%%M6401"
# dbname ="GOBP"; gsname = "GOBP_CYTOKINE_MEDIATED_SIGNALING_PATHWAY%%M898"  # up
dbname ="GOBP"; gsname = "GOBP_INFLAMMATORY_RESPONSE%%M10617"
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_CELL_ADHESION%%M8154"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_LIPID%%M14037"  # up&down
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_DRUG%%M6886"
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_INTERLEUKIN_1_MEDIATED_SIGNALING_PATHWAY%%M25511"
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_LIPID_LOCALIZATION%%M25391"  # up&down
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_CYTOKINE%%M15783"
dbname ="GOBP"; gsname = "GOBP_LIPID_LOCALIZATION%%M15749"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_MAPK_CASCADE%%M10496"   #!!! rn7
tgt = "hepato_m168101"; bgd = "hepato_sp"
#!!! rn7 ALL
dbname ="KEGG"; gsname = "KEGG_PPAR_SIGNALING_PATHWAY%%M13088"
dbname ="KEGG"; gsname = "KEGG_STEROID_BIOSYNTHESIS%%M5872"
dbname ="KEGG"; gsname = "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS%%M3214"
dbname ="REACTOME"; gsname = "REACTOME_CHOLESTEROL_BIOSYNTHESIS%%M16227"
dbname ="REACTOME"; gsname = "REACTOME_REGULATION_OF_CHOLESTEROL_BIOSYNTHESIS_BY_SREBP_SREBF%%M27001"
dbname ="REACTOME"; gsname = "REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_DEATH_RECEPTORS_AND_LIGANDS%%M27631" # only 10 genes
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_LIPID%%M14037"
dbname ="GOBP"; gsname = "GOBP_LIPID_STORAGE%%M15541"
dbname ="GOBP"; gsname = "GOBP_ORGANIC_ACID_METABOLIC_PROCESS%%M14117"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_DRUG%%M6886"
dbname ="GOBP"; gsname = "GOBP_ORGANIC_ACID_BIOSYNTHETIC_PROCESS%%M11689"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_MAPK_CASCADE%%M10496"
dbname ="GOBP"; gsname = "GOBP_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY%%M12608"

tgt = "stellate_m154207"; bgd = "stellate_sp"
# dbname ="KEGG"; gsname = "KEGG_LYSINE_DEGRADATION%%M13720"
# dbname ="KEGG"; gsname = "KEGG_STEROID_BIOSYNTHESIS%%M5872"   # rno00100 down
# dbname ="KEGG"; gsname = "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS%%M13465"
dbname ="REACTOME"; gsname = "REACTOME_CHOLESTEROL_BIOSYNTHESIS%%M16227"
dbname ="REACTOME"; gsname = "REACTOME_INTERLEUKIN_10_SIGNALING%%M27605"   #!!! rn7
# dbname ="WIKIPATHWAYS"; gsname = "WP_CHOLESTEROL_BIOSYNTHESIS_PATHWAY%%M39608"
tgt = "stellate_m167203"; bgd = "stellate_sp"
dbname ="GOBP"; gsname = "GOBP_ACTIN_FILAMENT_BASED_PROCESS%%M9082"  #up&down
dbname ="GOBP"; gsname = "GOBP_CELL_MIGRATION%%M40389"  #up&down
# dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_CELLULAR_COMPONENT_MOVEMENT%%M15511"
# dbname ="GOBP"; gsname = "GOBP_ENDOTHELIAL_CELL_FATE_COMMITMENT%%M24225"
# dbname ="GOBP"; gsname = "GOBP_EPITHELIAL_CELL_FATE_COMMITMENT%%M16390"  # up&down
tgt = "stellate_m168101"; bgd = "stellate_sp"
# none

tgt = "endothelial_m154207"; bgd = "endothelial_sp"
# dbname ="NetPath"; gsname = "IL2%NETPATH%IL2"
dbname ="NetPath"; gsname = "IL4%NETPATH%IL4"   #!!! rn7
tgt = "endothelial_m167203"; bgd = "endothelial_sp"
# dbname ="NetPath"; gsname = "WNT%NETPATH%WNT"
# dbname ="NetPath"; gsname = "ALPHA6BETA4INTEGRIN%NETPATH%ALPHA6BETA4INTEGRIN"
# dbname ="GOBP"; gsname = "GOBP_STEROID_METABOLIC_PROCESS%%M16758"  # down
dbname ="H"; gsname = "HALLMARK_INFLAMMATORY_RESPONSE%%M5932"   #!!! rn7
dbname ="REACTOME"; gsname = "REACTOME_MET_ACTIVATES_PI3K_AKT_SIGNALING%%M27741"   #!!! rn7; only 6 genes
tgt = "endothelial_m168101"; bgd = "endothelial_sp"
dbname ="KEGG"; gsname = "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION%%M9809"   #!!! rn7
dbname ="GOBP"; gsname = "GOBP_INFLAMMATORY_RESPONSE%%M10617"   #!!! rn7

tgt = "myelo_m154207";  bgd = "myelo_sp"
# dbname ="REACTOME"; gsname = "REACTOME_COMPLEMENT_CASCADE%%M19752"  # 0.04894015 irrelevant?
# dbname ="GOBP"; gsname = "GOBP_ALPHA_BETA_T_CELL_DIFFERENTIATION%%M10942"  # irrelevant?
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_IMMUNE_SYSTEM_PROCESS%%M5017"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_CELL_CELL_ADHESION%%M6401"   #!!! rn7
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_MAPK_CASCADE%%M10496"   #!!! rn7
tgt = "myelo_m167203";  bgd = "myelo_sp"
# dbname ="GOBP"; gsname = "GOBP_CELLULAR_LIPID_METABOLIC_PROCESS%%M14398"  # up&down
dbname ="GOBP"; gsname = "GOBP_LIPID_METABOLIC_PROCESS%%M14099"   #!!! rn7
tgt = "myelo_m168101";  bgd = "myelo_sp"
#!!! rn7 ALL
dbname ="GOBP"; gsname = "GOBP_ADAPTIVE_IMMUNE_RESPONSE%%M13847"
dbname ="GOBP"; gsname = "GOBP_POSITIVE_REGULATION_OF_LEUKOCYTE_CELL_CELL_ADHESION%%M25133"
dbname ="GOBP"; gsname = "GOBP_POSITIVE_REGULATION_OF_NATURAL_KILLER_CELL_MEDIATED_IMMUNITY%%M16164"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_T_CELL_ACTIVATION%%M16253"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_MAPK_CASCADE%%M10496"

tgt = "B_m154207";  bgd = "B_sp"
# none
tgt = "B_m167203";  bgd = "B_sp"
dbname ="H"; gsname = "HALLMARK_TNFA_SIGNALING_VIA_NFKB%%M5890"   #!!! rn7
tgt = "B_m168101";  bgd = "B_sp"
# none

tgt = "T_NK_m154207";  bgd = "T_NK_sp"
dbname ="GOBP"; gsname = "GOBP_FATTY_ACID_METABOLIC_PROCESS%%M17829"
# dbname ="H"; gsname = "HALLMARK_XENOBIOTIC_METABOLISM%%M5934"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_LIPID%%M14037"   #!!! rn7
tgt = "T_NK_m167203";  bgd = "T_NK_sp"
dbname ="GOBP"; gsname = "GOBP_STEROID_METABOLIC_PROCESS%%M16758"
dbname ="GOBP"; gsname = "GOBP_ORGANIC_ACID_METABOLIC_PROCESS%%M14117"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_DRUG%%M6886"
tgt = "T_NK_m168101";  bgd = "T_NK_sp"
# none
```

Not significant but important

```{r}
dbname ="WIKIPATHWAYS"; gsname = "WP_NONALCOHOLIC_FATTY_LIVER_DISEASE%%M39806"
#
dbname ="KEGG"; gsname = "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION%%M9809"
#
dbname ="REACTOME"; gsname = "REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION%%M610"
dbname ="REACTOME"; gsname = "REACTOME_DEGRADATION_OF_THE_EXTRACELLULAR_MATRIX%%M587"
dbname ="GOBP"; gsname = "GOBP_CONNECTIVE_TISSUE_REPLACEMENT%%M24786"
dbname ="REACTOME"; gsname = "REACTOME_COLLAGEN_FORMATION%%M631"
dbname ="REACTOME"; gsname = "REACTOME_COLLAGEN_DEGRADATION%%M26953"
dbname ="REACTOME"; gsname = "REACTOME_ACTIVATION_OF_MATRIX_METALLOPROTEINASES%%M26976"
#
dbname ="KEGG"; gsname = "KEGG_TGF_BETA_SIGNALING_PATHWAY%%M2642"  # rno04350
dbname ="WIKIPATHWAYS"; gsname = "WP_TGFBETA_SIGNALING_PATHWAY%%M39432"
dbname ="Panther"; gsname = "TGF-BETA SIGNALING PATHWAY%PANTHER PATHWAY%P00052"
#
dbname ="WIKIPATHWAYS"; gsname = "WP_FERROPTOSIS%%M39768"
dbname ="REACTOME"; gsname = "REACTOME_PYROPTOSIS%%M41804"  # FDR >0.9
dbname ="REACTOME"; gsname = "REACTOME_INTRINSIC_PATHWAY_FOR_APOPTOSIS%%M7455"  # FDR >0.8
dbname ="REACTOME"; gsname = "REACTOME_SUPPRESSION_OF_APOPTOSIS%%M29815"  # FDR >0.7
dbname ="GOBP"; gsname = "GOBP_APOPTOTIC_PROCESS%%M34075"
```

```{r}
gname = "Clusters4res05"

tgt = "C9";  bgd = "C8"
dbname ="GOBP"; gsname = "GOBP_CELL_MIGRATION%%M40389"
dbname ="GOBP"; gsname = "GOBP_CELL_CELL_ADHESION%%M12087"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_WOUNDING%%M5634"  # up&down
dbname ="GOBP"; gsname = "GOBP_LEUKOCYTE_CELL_CELL_ADHESION%%M14645"
dbname ="GOBP"; gsname = "GOBP_LEUKOCYTE_MIGRATION%%M12269"  # up&down
tgt = "C14";  bgd = "C13"
dbname ="H"; gsname = "HALLMARK_TNFA_SIGNALING_VIA_NFKB%%M5890"  # up&down
dbname ="KEGG"; gsname = "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION%%M9809"  # mostly up
dbname ="KEGG"; gsname = "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION%%M2164"  # up&down
dbname ="GOBP"; gsname = "GOBP_CELL_MIGRATION%%M40389"
dbname ="GOBP"; gsname = "GOBP_REGULATION_OF_CELL_ADHESION%%M8154"
dbname ="GOBP"; gsname = "GOBP_BLOOD_VESSEL_MORPHOGENESIS%%M13415"  # up&down
dbname ="GOBP"; gsname = "GOBP_POSITIVE_REGULATION_OF_REACTIVE_OXYGEN_SPECIES_METABOLIC_PROCESS%%M13580"  # up&down
dbname ="GOBP"; gsname = "GOBP_INFLAMMATORY_RESPONSE%%M10617"  # mostly up
dbname ="GOBP"; gsname = "GOBP_CYTOKINE_PRODUCTION%%M13156"
dbname ="GOBP"; gsname = "GOBP_RESPONSE_TO_WOUNDING%%M5634"  # up&down
```
