---
title: "ArchR"
output: html_notebook
---

Code snippets from ArchR tutorial

## 7.3 Identifying Marker Genes

```{r}
markersGS <- getMarkerFeatures(
    ArchRProj = proj2, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters2", # "Clusters",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    testMethod = "wilcoxon"
)

markerList <- getMarkers(
  markersGS,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25")
markerList$C6

markerGenes = c(
  "Alb", "Apoc3", "Cyp2e1", "Cyp2f4",
  "Epcam", "Krt19",
  "Dcn", "Des", "Lrat",
  "Stab2",
  "Csf3r",
  "Elane", "Ly6g6c", "Ly6g6d", "Ly6g6e", "Mpo",
  "Csf1r",
  "Lyz2",
  "Marco",
  "Ebf1",
  "Cd3g", "Gata3",
  "Gzma", "Prf1"
)
heatmapGS <- plotMarkerHeatmap(
  seMarker = markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  labelMarkers = markerGenes,
  transpose = TRUE
)
ComplexHeatmap::draw(heatmapGS, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```

## 7.4 Visualizing Marker Genes on an Embedding

```{r}
p <- plotEmbedding(
    ArchRProj = proj2, 
    colorBy = "GeneScoreMatrix", 
    name = markerGenes, 
    embedding = "UMAP",
    quantCut = c(0.01, 0.95),
    imputeWeights = NULL
)
p2 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p2))
plotPDF(plotList = p, 
    name = "Plot-UMAP-Marker-Genes-WO-Imputation.pdf", 
    ArchRProj = proj2, 
    addDOC = FALSE, width = 5, height = 5)
```

## 7.5 Marker Genes Imputation with MAGIC

```{r}
proj2 <- addImputeWeights(
  proj2,
  reducedDims = "myLSI2") # "myLSI"

p <- plotEmbedding(
    ArchRProj = proj2, 
    colorBy = "GeneScoreMatrix", 
    name = markerGenes, 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj2)
)
p2 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p2))
plotPDF(plotList = p, 
    name = "Plot-UMAP-Marker-Genes-W-Imputation.pdf", 
    ArchRProj = proj2, 
    addDOC = FALSE, width = 5, height = 5)
```

## 7.6 Track Plotting with ArchRBrowser

```{r}
p <- plotBrowserTrack(
    # ArchRProj = proj2, 
    # groupBy = "Clusters2", # "Clusters"
    ArchRProj = proj6, 
    groupBy = "Clusters4",
    geneSymbol = markerGenes, 
    upstream = 50000,
    downstream = 50000
)

# grid::grid.newpage()
# grid::grid.draw(p$Cyp2e1)

plotPDF(plotList = p, 
    name = "Plot-Tracks-Marker-Genes.pdf", 
    # ArchRProj = proj2, 
    ArchRProj = proj6, 
    addDOC = FALSE, width = 5, height = 5)
```

B-cell (Ebf1) vs T-cell (Cd3g, Gata3) / Tc & NK (Gzma, Prf1)
separate in Clusters2?
Yes: adjvar03 adjvar04
No:  150Ktopvar adjvar02 adjvar05

## 7.7 Launching the ArchRBrowser

```{r}
ArchRBrowser(proj2)
```

## 11.1 Identifying Marker Peaks with ArchR

Marker features are features that are unique to a specific cell grouping. These can be very useful in understanding cluster- or cell type-specific biology.

```{r}
markersPeaks <- getMarkerFeatures(
  ArchRProj = proj5, 
  useMatrix = "PeakMatrix", 
  # groupBy = "Clusters4res05", # "Clusters2",
  groupBy = "Clusters4BTmyelodiet",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)
markersPeaks

markerList <- getMarkers(markersPeaks, cutOff = "FDR <= 0.01 & Log2FC >= 1")
markerList
markerList$C12

markerList <- getMarkers(markersPeaks, cutOff = "FDR <= 0.01 & Log2FC >= 1", returnGR = TRUE)
markerList
markerList$C12
```

C9
chr18   79375052-79375552  Mbp
chr10   16942316-16942816
chr13   89959551-89960051  Cd244a
 chr3 164487649-164488149  Ptpn1

```{r}
write.table(
  markerList$C9,
  paste0("cluster.C9_DAR.FDRle001_Log2FCge1.genes.txt"),
  row.names = FALSE,
  col.names = TRUE,
  quote = FALSE,
  sep = "\t")

tgt = "C8"
Z = sign(assay(markersPeaks, "Log2FC", withDimnames = TRUE)) *
  qnorm(
    as.matrix(assay(markersPeaks, "Pval", withDimnames = TRUE) / 2),
    lower.tail = FALSE)
output = data.frame(
  name = rowData(markersPeaks)$name,
  Z = Z[[tgt]])
write.table(
  output,
  paste0("cluster.", tgt, "_genes.Z.rnk"),
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE,
  sep = "\t")
```

## 11.2 Plotting Marker Peaks in ArchR
### 11.2.1 Marker Peak Heatmaps

```{r}
heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markersPeaks, 
  cutOff = "FDR <= 0.1 & Log2FC >= 0.5",
  transpose = TRUE
)

draw(heatmapPeaks, heatmap_legend_side = "bot", annotation_legend_side = "bot")

plotPDF(heatmapPeaks,
        name = "Peak-Marker-Heatmap",
        width = 8, height = 6,
        ArchRProj = proj5, addDOC = FALSE)
```

### 11.2.2 Marker Peak MA and Volcano Plots

```{r}
pma <- plotMarkers(seMarker = markersPeaks, name = "C9", cutOff = "FDR <= 0.1 & Log2FC >= 1", plotAs = "MA")
pma

pv <- plotMarkers(seMarker = markersPeaks, name = "C9", cutOff = "FDR <= 0.1 & Log2FC >= 1", plotAs = "Volcano")
pv

# plotPDF(pma, pv, name = "Erythroid-Markers-MA-Volcano", width = 5, height = 5, ArchRProj = projHeme5, addDOC = FALSE)
```

### 11.2.3 Marker Peaks in Browser Tracks

```{r}
p <- plotBrowserTrack(
    ArchRProj = proj5, 
    groupBy = "Clusters4res05", # "Clusters2", 
    geneSymbol = c("Nlrp3"),
    features =  getMarkers(markersPeaks, cutOff = "FDR <= 0.1 & Log2FC >= 1", returnGR = TRUE)["C12"],
    upstream = 50000,
    downstream = 50000
)

grid::grid.newpage()
grid::grid.draw(p$Nlrp3)

# plotPDF(p, name = "Plot-Tracks-With-Features", width = 5, height = 5, ArchRProj = projHeme5, addDOC = FALSE)
```

## 11.3 Pairwise Testing Between Groups

```{r}
markerTest <- getMarkerFeatures(
  ArchRProj = proj5, 
  useMatrix = "PeakMatrix",
  # groupBy = "Clusters4res05", # "Clusters2",
  groupBy = "Clusters4BTmyelodiet",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  # useGroups = "C9",
  # bgdGroups = "C8",
  useGroups = "myelo_HFC",
  bgdGroups = "myelo_sp",
)
getMarkers(
  markerTest,
  cutOff = "FDR <= 0.01 & Log2FC >= 1", returnGR = TRUE)$C9

x = lapply(
  paste0("C", c(10:11, 13:17)),
  function (bgd) {
    mt <- getMarkerFeatures(
      ArchRProj = proj5, 
      useMatrix = "PeakMatrix",
      groupBy = "Clusters2",
      testMethod = "wilcoxon",
      bias = c("TSSEnrichment", "log10(nFrags)"),
      useGroups = "C12",
      bgdGroups = bgd,
    )
    gr = getMarkers(
      mt,
      cutOff = "FDR <= 0.01 & Log2FC >= 1", returnGR = TRUE)
    ov = GenomicRanges::findOverlaps(
      markerList$C12,
      gr$C12,
      select = "first")
    return(! is.na(ov))
  }
)
x = (do.call(pmin, x) > 0)
markerList$C12[x, ]
```

Clusters4res05, C9 vs C8
FDR <= 0.01 & Log2FC >= 1
chr1 198528203-198528703  LOC308990
chr5 127405596-127406096  Lrp8
chr7 143848930-143849430  Rarg
chr8   75686749-75687249  Anxa2

FDR <= 0.01 & Log2FC <= 1
chr10   61891373-61891873  Smg6
chr16   71295273-71295773  Fgfr1
chr10   31122394-31122894  Sox30 Adam19


Basic cell functional genes are detected.
Maybe the search conditions are inappropriate
TODO: bgd C10 C11 C13-C17
TODO: wider signal
TODO: GREAT

GRanges object with 235 ranges and 3 metadata columns:
        seqnames              ranges strand |    Log2FC         FDR  MeanDiff
           <Rle>           <IRanges>  <Rle> | <numeric>   <numeric> <numeric>
    [1]    chr19       932332-932832      * |   4.52336 7.04219e-23  0.317675
Tk2
    [2]     chr9   16425725-16426225      * |   4.62644 3.09141e-22  0.287891
Tbcc
    [3]     chr2 236509400-236509900      * |   4.33104 9.55290e-21  0.251414
    [4]     chr4   67056277-67056777      * |   4.75781 6.02731e-19  0.276610
    [5]    chr13   48531218-48531718      * |   4.38428 7.34858e-18  0.267285


GRanges object with 384 ranges and 3 metadata columns:
        seqnames              ranges strand |    Log2FC         FDR
           <Rle>           <IRanges>  <Rle> | <numeric>   <numeric>
    [1]     chr5 154519055-154519555      * |   2.88047 4.38479e-10
E2F2
    [2]     chr1 161954647-161955147      * |   4.75316 6.49054e-10
NARS2
    [3]    chr10   61361083-61361583      * |   3.90936 2.32291e-09
RAP1GAP2 (orthologue)
regulates secretion of dense granules from platelets at sites of endothelial damage
    [4]    chr10   37287700-37288200      * |   4.15956 1.77388e-08
Jade2
    [5]    chr14     3569284-3569784      * |   3.84262 3.28218e-08
    ...      ...                 ...    ... .       ...         ...
  [380]     chr4   67157874-67158374      * |   1.35424  0.00978440
  [381]     chr8   97462911-97463411      * |   2.28179  0.00978565
  [382]     chr8   90821175-90821675      * |   2.34380  0.00979108
  [383]     chr1 231363411-231363911      * |   2.13535  0.00981145
  [384]     chr3   93850062-93850562      * |   2.13073  0.00994300


```{r}
pma <- plotMarkers(seMarker = markerTest, name = "C12", cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1", plotAs = "MA")
pma

pv <- plotMarkers(seMarker = markerTest, name = "C12", cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1", plotAs = "Volcano")
pv

# plotPDF(pma, pv, name = "Erythroid-vs-Progenitor-Markers-MA-Volcano", width = 5, height = 5, ArchRProj = projHeme5, addDOC = FALSE)
```

### Export Differential Peaks

```{r}
identical(rowData(markersPeaks), rowData(markerTest))

tgt = "C9" # target
bgd = "C8" # background
tgt = "myelo_HFC" # target
bgd = "myelo_sp" # background
plotdata =
  sign(assay(markersPeaks, "Log2FC", withDimnames = TRUE)[, c(tgt, bgd)]) *
  qnorm(
    as.matrix(assay(markersPeaks, "Pval", withDimnames = TRUE)[, c(tgt, bgd)] / 2),
    lower.tail = FALSE)
diffaccess =
  sign(assay(markerTest, "Log2FC", withDimnames = TRUE)) *
  qnorm(
    as.matrix(assay(markerTest, "Pval", withDimnames = TRUE) / 2),
    lower.tail = FALSE)
colnames(diffaccess) = "diffaccess"
plotdata = cbind(plotdata, diffaccess)
# For GeneScoreMatrix
plotdata$name = rowData(markersPeaks)$name

ggplot(
  data = plotdata,
  aes(x = .data[[tgt]],
      # y = .data[[bgd]]
      y = diffaccess
      )) +
  geom_bin2d() +
  scale_fill_gradient(name = "count", trans = "log10") +
  labs(x = paste0("Z-score for ", tgt),
       # y = paste0("Z-score for ", bgd),
       y = "Z-score for diffaccess",
       title = "Accessibility of peaks")

x = 
  (assay(markersPeaks, "Log2FC", withDimnames = TRUE)[, tgt] > 1) &
  (assay(markersPeaks, "FDR", withDimnames = TRUE)[, tgt] < 0.01) &
  (assay(markerTest, "Log2FC", withDimnames = TRUE)[, 1] > 1) &
  (assay(markerTest, "FDR", withDimnames = TRUE)[, 1] < 0.01) 

xx = rowData(markersPeaks)[x, ]

# For GeneScoreMatrix
write.table(
  xx$name,
  paste0("cluster.", tgt, "vs", bgd, "_DAR.genes.txt"),
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE)

# For PeakMatrix
xx = xx[, c("seqnames", "start", "end")]
xx$name = paste0(xx$seqnames, ":", xx$start, "-", xx$end)
xx$score = 0
xx$strand = "."
xx$start = xx$start - 1
write.table(
  xx,
  "cluster.foo_DAR.bed",
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE,
  sep = "\t"
)



Z = sign(assay(markerTest, "Log2FC", withDimnames = TRUE)) *
  qnorm(
    as.matrix(assay(markerTest, "Pval", withDimnames = TRUE) / 2),
    lower.tail = FALSE)
output = data.frame(
  name = rowData(markerTest)$name,
  Z = Z[[tgt]])
write.table(
  output,
  paste0("cluster.", tgt, "vs", bgd, "_genes.Z.rnk"),
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE,
  sep = "\t")
```

script_CrossMap

Use website http://great.stanford.edu/public/html/

## 12.1 Motif Enrichment in Differential Peaks

14.926 mins

```{r}
proj5 <- addMotifAnnotations(
  ArchRProj = proj5,
  motifSet = "cisbp",
  species = "Mus musculus", # Homo sapiens",
  name = "Motif")

proj5@peakAnnotation@listData$Motif$motifSummary
```

`markerTest` was defined above

```{r}
motifsUp <- peakAnnoEnrichment(
    seMarker = markerTest,
    ArchRProj = proj5,
    peakAnnotation = "Motif",
    cutOff = "FDR <= 0.1 & Log2FC >= 0.5"
  )

motifsUp

df <- data.frame(
  TF = rownames(motifsUp),
  mlog10Padj = assay(motifsUp)[,1])
df <- df[order(df$mlog10Padj, decreasing = TRUE),]
df$rank <- seq_len(nrow(df))
head(df)
```
```{r}
ggUp <- ggplot(df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) +
  ggrepel::geom_label_repel(
        data = df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,
        nudge_x = 2,
        color = "black"
  ) + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + 
  xlab("Rank Sorted TFs Enriched") +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))

ggUp
```

```{r}
motifsDo <- peakAnnoEnrichment(
    seMarker = markerTest,
    ArchRProj = proj5,
    peakAnnotation = "Motif",
    cutOff = "FDR <= 0.1 & Log2FC <= -0.5"
  )

motifsDo

df <- data.frame(
  TF = rownames(motifsDo),
  mlog10Padj = assay(motifsDo)[,1])
df <- df[order(df$mlog10Padj, decreasing = TRUE),]
df$rank <- seq_len(nrow(df))
head(df)
```
```{r}
ggDo <- ggplot(df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) +
  ggrepel::geom_label_repel(
        data = df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,
        nudge_x = 2,
        color = "black"
  ) + theme_ArchR() + 
  ylab("-log10(FDR) Motif Enrichment") +
  xlab("Rank Sorted TFs Enriched") +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))

ggDo
```

```{r}
plotPDF(ggUp, ggDo, name = "C12-vs-C11-Markers-Motifs-Enriched", width = 5, height = 5, ArchRProj = proj5, addDOC = FALSE)
```

## 12.2 Motif Enrichment in Marker Peaks

`markersPeaks` was defined above

```{r}
enrichMotifs <- peakAnnoEnrichment(
    seMarker = markersPeaks,
    ArchRProj = proj5,
    peakAnnotation = "Motif",
    cutOff = "FDR <= 0.1 & Log2FC >= 0.5"
  )

enrichMotifs
```

```{r}
heatmapEM <- plotEnrichHeatmap(enrichMotifs, n = 3, transpose = TRUE)

ComplexHeatmap::draw(heatmapEM, heatmap_legend_side = "bot", annotation_legend_side = "bot")

plotPDF(heatmapEM, name = "Motifs-Enriched-Marker-Heatmap", width = 8, height = 6, ArchRProj = proj5, addDOC = FALSE)
```

# 13 ChromVAR Deviatons Enrichment with ArchR

A deviation is a bias-corrected measurement of how far the per-cell accessibility of a given feature (i.e motif) deviates from the expected accessibility based on the average of all cells or samples.

## 13.1 Motif Deviations

```{r}
if("Motif" %ni% names(proj5@peakAnnotation)){
    proj5 <- addMotifAnnotations(
      ArchRProj = proj5,
      motifSet = "cisbp",
      name = "Motif")
}

proj5 <- addBgdPeaks(proj5)
```

This function has an optional parameter called matrixName that allows us to define the name of the deviations matrix that will be stored in the Arrow files. If we do not provide a value to this parameter, as in the example below, this function creates a matrix name by adding the word "Matrix" to the name of the peakAnnotation. The example below creates a deviations matrix in each of our Arrow files called "MotifMatrix".

48.483 mins

```{r}
proj5 <- addDeviationsMatrix(
  ArchRProj = proj5, 
  peakAnnotation = "Motif",
  force = TRUE
)
```

rank the variability of the deviations

```{r}
plotVarDev <- getVarDeviations(proj5, name = "MotifMatrix", plot = TRUE)

plotVarDev

plotPDF(plotVarDev, name = "Variable-Motif-Deviation-Scores", width = 5, height = 5, ArchRProj = proj5, addDOC = FALSE)
```

extract a subset of motifs for downstream analysis

```{r}
motifs <- c("BCL11A", "HNF4A", "HNF4G", "SPI1", "SPIB", "ERG")
markerMotifs <- getFeatures(proj5, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")
markerMotifs

markerMotifs <- grep("z:", markerMotifs, value = TRUE)
markerMotifs
```

```{r}
p <- plotGroups(ArchRProj = proj5, 
  groupBy = "Clusters2", 
  colorBy = "MotifMatrix", 
  name = markerMotifs,
  imputeWeights = getImputeWeights(proj5)
)

p2 <- lapply(seq_along(p), function(x){
  if(x != 1){
    p[[x]] + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6) +
    theme(plot.margin = unit(c(0.1, 0.1, 0.1, 0.1), "cm")) +
    theme(
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank(),
        axis.title.y=element_blank()
    ) + ylab("")
  }else{
    p[[x]] + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6) +
    theme(plot.margin = unit(c(0.1, 0.1, 0.1, 0.1), "cm")) +
    theme(
        axis.ticks.y=element_blank(),
        axis.title.y=element_blank()
    ) + ylab("")
  }
})
do.call(cowplot::plot_grid, c(list(nrow = 1, rel_widths = c(2, rep(1, length(p2) - 1))),p2))

plotPDF(p, name = "Plot-Groups-Deviations-w-Imputation", width = 5, height = 5, ArchRProj = proj5, addDOC = FALSE)
```

```{r}
p <- plotEmbedding(
    ArchRProj = proj5, 
    colorBy = "MotifMatrix", 
    name = sort(markerMotifs), 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj5)
)

p2 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p2))

# Plot-UMAP-markerMotifs-MotifMatrix.pdf
```

To see how these TF deviation z-scores compare to the inferred gene expression via gene scores of the corresponding TF genes, we can overlay the gene scores for each of these TFs on the UMAP embedding.

Differs somewhat from MotifMatrix !!
MotifMatrix z:HNF4A_662 differs among hepatocytes in Clusters2
GeneScoreMatrix Hnf4a is uniform across hepatocytes

```{r}
markerRNA <- getFeatures(proj5, select = paste(motifs, collapse="|"), useMatrix = "GeneScoreMatrix")
markerRNA <- markerRNA[markerRNA %ni% c("Catsperg","Tcerg1l", "Tcerg1", "Ergic3", "Rergl", "Ergic2", "Erg28" )]
markerRNA
```

```{r}
p <- plotEmbedding(
    ArchRProj = proj5, 
    colorBy = "GeneScoreMatrix", 
    name = sort(markerRNA), 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj5)
)

p2 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p2))

# Plot-UMAP-markerMotifs-GeneScoreMatrix.pdf
```

## 14.1 Motif Footprinting

the first thing we need to do is obtain the positions of the relevant motifs. To do this, we call the getPositions() function. This function has an optional parameter called name which can accept the name of the peakAnnotation object from which we would like to obtain the positions. If name = NULL, then ArchR will use the first entry in the peakAnnotation slot. In the example shown below, we do not specify name and ArchR uses the first entry which is our CIS-BP motifs.

```{r}
motifPositions <- getPositions(proj5)
motifPositions
```

```{r}
motifs <- c("BCL11A", "HNF4A", "HNF4G", "SPI1", "SPIB", "ERG")
markerMotifs <- unlist(lapply(motifs, function(x) grep(x, names(motifPositions), value = TRUE)))
markerMotifs
```

To accurately profile TF footprints, a large number of reads are required. Therefore, cells are grouped to create pseudo-bulk ATAC-seq profiles that can be then used for TF footprinting. These pseudo-bulk profiles are stored as group coverage files which we originally created in a previous chapter to perform peak calling.

```{r}
seFoot <- getFootprints(
  ArchRProj = proj5, 
  positions = motifPositions[markerMotifs], 
  groupBy = "Clusters2"
)
```

## 14.2 Normalization of Footprints for Tn5 Bias
### 14.2.1 Subtracting the Tn5 Bias

```{r}
plotFootprints(
  seFoot = seFoot,
  ArchRProj = proj5, 
  normMethod = "Subtract",
  plotName = "Footprints-Subtract-Bias",
  addDOC = FALSE,
  smoothWindow = 5
)
```

### 14.2.2 Dividing by the Tn5 Bias

```{r}
plotFootprints(
  seFoot = seFoot,
  ArchRProj = proj5, 
  normMethod = "Divide",
  plotName = "Footprints-Divide-Bias",
  addDOC = FALSE,
  smoothWindow = 5
)
```

## 14.3 Feature Footprinting

A TSS insertion profile is just a specialized sub-case of footprinting.

```{r}
seTSS <- getFootprints(
  ArchRProj = proj5, 
  positions = GRangesList(TSS = getTSS(proj5)), 
  groupBy = "Clusters2",
  flank = 2000
)
```

```{r}
plotFootprints(
  seFoot = seTSS,
  ArchRProj = proj5, 
  normMethod = "None",
  plotName = "TSS-No-Normalization",
  addDOC = FALSE,
  flank = 2000,
  flankNorm = 100
)
```

## 15.2 Co-accessibility with ArchR

create low-overlapping aggregates of single cells prior to these analyses

One thing to note about co-accessibility analysis is that it often identified cell type-specific peaks as being co-accessibile. This is because these peaks are often all accessible together within a single cell type and often all not accessible in all other cell types. This drives a strong correlation but does not necessarily mean that there is a regulatory relationship between these peaks.

TODO: adjust by major cell-type ?? by Cluster2 ??

```{r}
proj5 <- addCoAccessibility(
    ArchRProj = proj5,
    reducedDims = "myLSI"
)
```

```{r}
cA <- getCoAccessibility(
    ArchRProj = proj5,
    corCutOff = 0.5,
    resolution = 1,
    returnLoops = FALSE
)
cA
```

The indexes of queryHits and subjectHits mentioned above apply to this GRanges object.

```{r}
metadata(cA)[[1]]
```

```{r}
cA <- getCoAccessibility(
    ArchRProj = proj5,
    corCutOff = 0.5,
    resolution = 1,
    returnLoops = TRUE
)
cA[[1]]
```

If, instead, we decrease the resolution of our loops to resolution = 1000, this can help with over-plotting of co-accessibility interactions.

```{r}
cA <- getCoAccessibility(
    ArchRProj = proj5,
    corCutOff = 0.5,
    resolution = 1000,
    returnLoops = TRUE
)
cA[[1]]
```

### 15.2.1 Plotting browser tracks of Co-accessibility

```{r}
markerGenes = c(
  "Cyp2e1",
  "Alb",
  "Dcn",
  "Stab2",
  "Csf1r",
  "Csf3r",
  "Ebf1",
  "Prf1",
  "Cd3g",
  
  "Nlrp3"
)
p <- plotBrowserTrack(
    ArchRProj = proj5, 
    groupBy = "Clusters2", 
    geneSymbol = markerGenes, 
    upstream = 50000,
    downstream = 50000,
    loops = getCoAccessibility(proj5)
)

grid::grid.newpage()
grid::grid.draw(p$Cyp2e1)

plotPDF(plotList = p, 
    name = "Plot-Tracks-Marker-Genes-with-CoAccessibility.pdf", 
    ArchRProj = proj5, 
    addDOC = FALSE, width = 5, height = 5)
```

## 15.4 Identification of Positive TF-Regulators

identify TFs whose gene expression is positively correlated to changes in the accessibility of their corresponding motif.  ArchR can identify TFs whose inferred gene scores are correlated to their chromVAR TF deviation z-scores.

low-overlapping cell aggregates

### 15.4.1 Step 1. Identify Deviant TF Motifs

```{r}
seGroupMotif <- getGroupSE(
  ArchRProj = proj5,
  useMatrix = "MotifMatrix",
  groupBy = "Clusters2")
```

Because this SummarizedExperiment object comes from the MotifMatrix is has two seqnames - "deviations" and "z" - corresponding to the raw deviations and deviation z-scores from chromVAR.

```{r}
seGroupMotif
head(rowData(seGroupMotif))
tail(rowData(seGroupMotif))
```

We can subset this SummarizedExperiment to just the deviation z-scores.

```{r}
seZ <- seGroupMotif[rowData(seGroupMotif)$seqnames=="z",]
```

Then we can identify the maximum delta in z-score between all clusters. This will be helpful in stratifying motifs based on the degree of variation observed across clusters.

```{r}
rowData(seZ)$maxDelta <-
  lapply(
    seq_len(ncol(seZ)),
    function(x){
      rowMaxs(assay(seZ) - assay(seZ)[,x])
    }) %>%
  Reduce("cbind", .) %>%
  rowMaxs
```

### 15.4.2 Step 2. Identify Correlated TF Motifs and TF Gene Score/Expression

To identify TFs whose motif accessibility is correlated with with their own gene activity (either by gene score or gene expression), we use the correlateMatrices() function and provide the two matrices that we are interested in, in this case the GeneScoreMatrix and the MotifMatrix. As mentioned previously, these correlations are determined across many low-overlapping cell aggregates determined in the lower dimension space specified in the reducedDims parameter.

```{r}
corGSM_MM <- correlateMatrices(
    ArchRProj = proj5,
    useMatrix1 = "GeneScoreMatrix",
    useMatrix2 = "MotifMatrix",
    reducedDims = "myLSI"
)

corGSM_MM
```

### 15.4.3 Step 3. Add Maximum Delta Deviation to the Correlation Data Frame

```{r}
corGSM_MM$maxDelta <- rowData(seZ)[match(corGSM_MM$MotifMatrix_name, rowData(seZ)$name), "maxDelta"]
```

### 15.4.4 Step 4. Identify Positive TF Regulators

we consider positive regulators as those TFs whose correlation between motif and gene score (or gene expression) is greater than 0.5 with an adjusted p-value less than 0.01 and a maximum inter-cluster difference in deviation z-score that is in the top quartile.

```{r}
corGSM_MM <- corGSM_MM[order(abs(corGSM_MM$cor), decreasing = TRUE), ]
corGSM_MM <- corGSM_MM[which(!duplicated(gsub("\\-.*","",corGSM_MM[,"MotifMatrix_name"]))), ]
corGSM_MM$TFRegulator <- "NO"
corGSM_MM$TFRegulator[which(corGSM_MM$cor > 0.5 & corGSM_MM$padj < 0.01 & corGSM_MM$maxDelta > quantile(corGSM_MM$maxDelta, 0.75))] <- "YES"
sort(corGSM_MM[corGSM_MM$TFRegulator=="YES",1])
```
```{r}
p <- ggplot(data.frame(corGSM_MM), aes(cor, maxDelta, color = TFRegulator)) +
  geom_point() + 
  theme_ArchR() +
  geom_vline(xintercept = 0, lty = "dashed") + 
  scale_color_manual(values = c("NO"="darkgrey", "YES"="firebrick3")) +
  xlab("Correlation To Gene Score") +
  ylab("Max TF Motif Delta") +
  scale_y_continuous(
    expand = c(0,0), 
    limits = c(0, max(corGSM_MM$maxDelta)*1.05)
  )

p
```
